File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/composite/Composite.java
package jv.com.tencent.cain.design_pattern.composite;

import java.util.ArrayList;
import java.util.List;

/**
 * @author cainjiang
 * @date 2020/5/7
 */
public class Composite extends Component {
    public static final String TAG = "Composite";

    List<Component> children = new ArrayList<>();

    public Composite() {
    }

    public Composite(String name) {
        super(name);
    }

    @Override
    public void add(Component component) {
        System.out.println(TAG + "#" + "add# " + "component: " + component.toString());
        children.add(component);
    }

    @Override
    public void remove(Component component) {
        if (children.remove(component)) {
            System.out.println(TAG + "#" + "remove# " + "component: " + component.toString());
        }
    }

    @Override
    public void display(int depth) {
        StringBuilder depthBuilder = new StringBuilder();
        for (int i = 0; i < depth; i++) {
            depthBuilder.append("-");
        }
        depthBuilder.append(name);
        System.out.println(depthBuilder);
        for (Component component : children) {
            component.display(depth + 1);
        }
    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/composite/Composite.kt
package kt.com.tencent.cain.desigin_pattern.composite

/**
 * @author cainjiang
 * @date 2020/5/7
 */
class Composite(name: String = "", val children: MutableList<Component> = mutableListOf()) : Component(name) {
    companion object {
        const val TAG = "Composite"
    }

    override fun add(component: Component) {
        println(TAG + "#add# component: $component")
        children.add(component)
    }

    override fun remove(component: Component) {
        if (children.remove(component)) {
            println(TAG + "#remove# component: $component")
        }
    }

    override fun display(depth: Int) {
        val depthBuilder = StringBuilder()
        for (i in 0 until depth) {
            depthBuilder.append("-")
        }
        depthBuilder.append(name)
        println(depthBuilder)
        children.forEach { it.display(depth + 1) }
    }
}